### PR TITLE
feat: [IDP-454]: Support Infra status for multiple pods

### PR DESCRIPTION
**Pull request by vigneswara-propelo**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/47358)

### CODE REVIEW

**Comment by vigneswara-propelo** [comment]
trigger uts

File: idp-service/src/main/java/io/harness/idp/k8s/client/K8sApiClient.java [hunk 1]
```diff
@@ -50,6 +51,7 @@ public class K8sApiClient implements K8sClient {
   @Inject @Named("backstageSaToken") private String backstageSaToken;
   @Inject @Named("backstageSaCaCrt") private String backstageSaCaCrt;
   @Inject @Named("backstageMasterUrl") private String backstageMasterUrl;
+  @Inject @Named("backstagePodLabel") private String backstagePodLabel;
```
Comment (sathish-soundarapandian) [inline idp-service/src/main/java/io/harness/idp/k8s/client/K8sApiClient.java:54]: nit: Better to have these configs related to backstage in a Class. Something like "BackstageConfig.java"

File: idp-service/src/main/java/io/harness/idp/k8s/client/K8sApiClient.java [hunk 2]
```diff
@@ -86,6 +88,24 @@ public V1ConfigMap updateConfigMapData(
...
+  @Override
+  public V1PodList getBackstagePodList(String namespace) {
+    KubernetesConfig kubernetesConfig = getKubernetesConfig();
+    ApiClient apiClient = kubernetesHelperService.getApiClient(kubernetesConfig);
+    CoreV1Api coreV1Api = new CoreV1Api(apiClient);
+    final Supplier<V1PodList> podListSupplier = Retry.decorateSupplier(retry, () -> {
+      try {
+        return coreV1Api.listNamespacedPod(
+            namespace, null, null, null, null, backstagePodLabel, null, null, null, null, false);
+      } catch (ApiException e) {
```
Comment (sathish-soundarapandian) [inline idp-service/src/main/java/io/harness/idp/k8s/client/K8sApiClient.java:100]: Any other exception that can be anticipated here ?

File: idp-service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java [hunk 3]
```diff
@@ -44,32 +37,29 @@ public class PodHealthCheck implements HealthCheck {
...
-          statusInfo.setReason(getPodMessage(pod));
+    String namespace = getNamespaceForAccountId(accountId);
+    V1PodList podList = k8sClient.getBackstagePodList(namespace);
+    if (CollectionUtils.isEmpty(podList.getItems())) {
+      statusInfo.setCurrentStatus(CurrentStatusEnum.NOT_FOUND);
+      statusInfo.setReason("No pod exists for namespace: " + namespace);
+    } else {
+      int failedPods = 0;
+      for (V1Pod pod : podList.getItems()) {
+        if (isPodInRunningPhase(pod) && isContainerInRunningState(pod) && isPodInReadyCondition(pod) != null) {
```
Comment (sathish-soundarapandian) [inline idp-service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java:48]: Question - So the logic is if one among N pods is in running phase, then we are all good. Considering this running pod will be able to serve requests. Looks ok, but do consider if there can be any corner cases based on the pod state.

File: idp-service/src/test/java/io/harness/idp/status/k8s/PodHealthCheckTest.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+    mockNamespaceInfo();
+  }
+
+  @Test
+  @Category(UnitTests.class)
+  public void testGetCurrentStatusWithNoPods() {
```
Comment (sathish-soundarapandian) [inline idp-service/src/test/java/io/harness/idp/status/k8s/PodHealthCheckTest.java:59]: nit: Please add owner annotation for tracking

File: idp-service/src/main/java/io/harness/idp/k8s/client/K8sApiClient.java (hunk 1 above)
Comment (vigneswara-propelo) [reply idp-service/src/main/java/io/harness/idp/k8s/client/K8sApiClient.java:54]: sure. Will take this up in separate PR as we have other backstage related configs which are used in other classes.

File: idp-service/src/main/java/io/harness/idp/k8s/client/K8sApiClient.java (hunk 2 above)
Comment (vigneswara-propelo) [reply idp-service/src/main/java/io/harness/idp/k8s/client/K8sApiClient.java:100]: No, ApiException is the only exception that we will get

File: idp-service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java (hunk 3 above)
Comment (vigneswara-propelo) [reply idp-service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java:48]: yes correct. During provisioning, if pod is in ready condition (isPodInReadyCondition check) it wont go back to waiting//terminated state. Also, for onboarding one pod is enough to serve the traffic

File: idp-service/src/test/java/io/harness/idp/status/k8s/PodHealthCheckTest.java (hunk 4 above)
Comment (vigneswara-propelo) [reply idp-service/src/test/java/io/harness/idp/status/k8s/PodHealthCheckTest.java:59]: done

**Review by sathish-soundarapandian — APPROVED** [review]

**Review by vikyathharekal — APPROVED** [review]
> Reviewed 2 of 4 files at r1, 1 of 3 files at r2, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47358)* status: 5 file(s) to review. 3. unresolved discussions.

**Comment by vikyathharekal** [comment]
trigger review

**Comment by vigneswara-propelo** [comment]
trigger uts
